### ISSUE TITLE
Kiali operator support HPA

### ISSUE THREAD

**Issue by zackzhangkai** [kiali/kiali#3533]
**Is your feature request related to a problem? Please describe.**

After install Kiali Operator , there are no HPA for kiali pod. Maybe it is better to support HPA.

**Comment by jshaughn** [comment]
@zackzhangkai It's not totally clear to me what you are expecting or asking for with this request.  Could you update the description with some more detail around what you want and how it is beneficial to Kiali users?  Thanks.

**Comment by zackzhangkai** [comment]
@jshaughn Thanks for your reply.
HPA: The Horizontal Pod Autoscaler automatically scales the number of Pods in a replication controller, deployment. 
It's appreciate that if the HPA is provided after install Kiali by Kiali operator.

**Comment by jmazzitelli** [comment]
I will investigate.

**Comment by jmazzitelli** [comment]
We could add a feature to support simple HPA creation.

So one simple proposal would be to have the operator create a `HorizontalPodAutoscaler` resource similar to how `kubectl autoscale` is documented and works: https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#autoscale ... so the Kiali CR could have `hpa.min`, `hpa.max`, `hpa.cpu-percent` settings in the `spec.deployment` section.

BUT! There are relatively new features within HPA that could also be specified (I'm thinking of behavior definitions and custom metrics). Who knows what k8s will add in the future. So... rather than encode the same complex yaml that HPA supports within our Kiali CR schema, we could simply add a generic `hpa.spec` section that allows the user to tell us exactly what they want in their Kiali HPA resource. This would support all current and future HPA settings. So this would be the proposal:

```
spec:
  deployment:
    hpa:
      api_version: autoscaling/v2beta2
      spec:
        ... your custom HPA yaml that will go in the HPA the operator will create ...
```

So, for example, the user could have this in the Kiali CR:

```
spec:
  deployment:
    hpa:
      api_version: autoscaling/v2beta2
      spec:
        maxReplicas: 5
        minReplicas: 2
        targetCPUUtilizationPercentage: 80
        behavior:
          scaleDown:
            stabilizationWindowSeconds: 10
        metrics:
        - type: Resource
          resource:
              name: cpu
              target:
                  type: Utilization
                  averageUtilization: 50
```

And everything under `hpa.spec` will go inside the created HPA resource's spec section. Note that the user won't have to specify (in fact, should not specify) the `scaleTargetRef` because the operator will create that, but the user can optionally say what apiVersion of HPA they want to use (this would support future HPA versions without the operator having to be updated). So the operator would create this:

```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: kiali
  namespace: istio-system
spec:
  maxReplicas: 5
  minReplicas: 2
  targetCPUUtilizationPercentage: 80
  behavior:
    scaleDown:
      stabilizationWindowSeconds: 10
  metrics:
  - type: Resource
    resource:
        name: cpu
        target:
            type: Utilization
            averageUtilization: 50
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: kiali
```

**Comment by jmazzitelli** [comment]
Still need to test these, but here's the work in progress that can implement this feature:

* kiali/kiali PR: https://github.com/kiali/kiali/pull/3542
* operator PR: https://github.com/kiali/kiali-operator/pull/230
* helm-charts PR: https://github.com/kiali/helm-charts/pull/26

**Comment by jmazzitelli** [comment]
@zackzhangkai What does your HPA resource look like? Can you attach an HPA resource that you want to use? I'll see if the PRs support what you want.

**Comment by zackzhangkai** [comment]
Thanks a lot.
Such as the jaeger, after I installed jaeger operator, then I would add a HPA configuration in the CR.

```bash
kubectl get jaeger jaeger  -oyaml
...
spec:
...
  collector:
    maxReplicas: 5
...

# kubectl get hpa jaeger-collector -oyaml
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  labels:
    app: jaeger
    app.kubernetes.io/component: hpa-collector
    app.kubernetes.io/instance: jaeger
    app.kubernetes.io/managed-by: jaeger-operator
    app.kubernetes.io/name: jaeger-collector
    app.kubernetes.io/part-of: jaeger
  name: jaeger-collector
  namespace: istio-system
  ownerReferences:
  - apiVersion: jaegertracing.io/v1
    controller: true
    kind: Jaeger
    name: jaeger
spec:
  maxReplicas: 5
  minReplicas: 1
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: jaeger-collector
  targetCPUUtilizationPercentage: 90
status:
  currentReplicas: 1
  desiredReplicas: 0
```

https://www.jaegertracing.io/docs/1.21/operator/

**Comment by jmazzitelli** [comment]
I ran a quick test with the PRs. In Kiali CR I had this:
```
spec:
  deployment:
    hpa:
      spec:
        maxReplicas: 3
        minReplicas: 2
        targetCPUUtilizationPercentage: 50
```
After the operator finished, I see this:
```
$ oc get horizontalpodautoscalers.autoscaling -n istio-system
NAME    REFERENCE          TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
kiali   Deployment/kiali   <unknown>/80%   2         3         2          109s

$ oc get pods -n istio-system -l app=kiali
NAME                     READY   STATUS    RESTARTS   AGE
kiali-6bf577cb5b-mflnj   1/1     Running   0          94s
kiali-6bf577cb5b-v9r6c   1/1     Running   0          112s
```

HPA that was created by the operator looked like this (I deleted some k8s-generated fields to make it easier to read):

```
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  labels:
    app: kiali
    app.kubernetes.io/name: kiali
    app.kubernetes.io/part-of: kiali
    app.kubernetes.io/version: dev
    version: dev
  name: kiali
  namespace: istio-system
spec:
  maxReplicas: 3
  minReplicas: 2
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: kiali
  targetCPUUtilizationPercentage: 80
status:
  currentReplicas: 2
  desiredReplicas: 2
  lastScaleTime: "2020-12-21T18:59:21Z"
```

I pointed my browser to the Kiali route URL and things worked - I was able to access the Kiali UI.

**Comment by jmazzitelli** [comment]
(notice the targetCPUUtilization didn't get set to 50 - I think its because my cluster doesn't support v2beta2. Notice the HPA that was created was v1, but the default HPA the operator tries to create is v2beta2. So I guess the cluster will downgrade the resource version if it doesn't support what you asked for).

**Comment by jmazzitelli** [comment]
I added a molecule test to the operator PR. The three PRs are ready for review and merge:

* kiali/kiali PR: https://github.com/kiali/kiali/pull/3542
* operator PR: https://github.com/kiali/kiali-operator/pull/230
* helm-charts PR: https://github.com/kiali/helm-charts/pull/26

**Comment by jmazzitelli** [comment]
To test, pull the three PRs, push the operator dev image to the cluster and run the molecule test where the new HPA stuff is and confirm it all passes:

```
make -e MOLECULE_USE_DEV_IMAGES=true -e MOLECULE_SCENARIO="affinity-tolerations-resources-test" make cluster-push molecule-test
```

**Comment by zackzhangkai** [comment]
Hi, how can I use the the HPA feature?  Does the latest release have supported now?

**Comment by jmazzitelli** [comment]
> Hi, how can I use the the HPA feature? Does the latest release have supported now?

This new feature is not available yet. It will be available in v1.29.0 which is due to be released in 3 days (should be available by sometime in the late afternoon/evening (US-Eastern time) on January 22, 2021.)
